service.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2019 RedHat
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"context"
+	"strconv"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	"github.com/operator-framework/operator-sdk/pkg/k8sutil"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var (
+	defaultMetricsPath = "/customMetrics"
+	defaultMetricsPort = "8089"
+)
+
+type convertmetrics func()
+
+type PrometheusUser struct {
+	Metricpath            string
+	Metricport            string
+	MetricsList           []prometheus.Collector
+	RecordMetricsFunction convertmetrics
```

Comment:
I'm trying to think of a use case for user defined metrics register function. But that doesn't mean there isn't one. Could you expand a bit more on including this?

Reply:
The user would have to provide the process in which the metrics defined in metricsList are to be collected. Or else, we wont be able to do anything having the  list of custom metrics we get. I assume, if the user provides MetricsList, then they should also provide RecordMetricsFunction. Both these parameters seem to be mutually inclusive as both of them are required to expose custom metrics.

Reply:
I guess I'm confused by what you mean by `collected`? What would a user want other than what's in `RegisterMetrics()`?

Reply:
I think the Collector interface does not include the function that provides the process in which the metrics are collected. For example: "RegisterMetrics" would register the values present in the list ([]prometheus.Collector) with prometheus. But it does not tell us on how those values are to be collected. For that purpose we would require RecordMetricsFunction.

Reply:
Ah, to update metrics I see. So really the user should be creating these functions and calling them in their code. Each metric may have its own update function, so at minimum this would need to be a list of functions.